browser-tests/event-search/eventSearchPage.components.ts

```diff
@@ -0,0 +1,279 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+import { screen, within } from '@testing-library/testcafe';
+import TestController from 'testcafe';
+
+import translations from '../../src/common/translation/i18n/fi.json';
+import { getEventFields } from '../../src/domain/event/EventUtils';
+import { KeywordOption } from '../../src/domain/event/types';
+import { formatDate } from '../../src/util/dateUtils';
+import getDateRangeStr from '../../src/util/getDateRangeStr';
+import toPascalCase from '../../src/util/toPascalCase';
+import { getErrorMessage } from '../utils/error.util';
+import { getExpectedEventContext } from '../utils/event.utils';
+import {
+  EventFieldsFragment,
+  Neighborhood,
+  PlaceFieldsFragment,
+} from '../utils/generated/graphql';
+import { regExpEscaped } from '../utils/regexp.util';
+
+export const getEventSearchPageComponents = (t: TestController) => {
+  const searchContainer = () => {
+    const selectors = {
+      component() {
+        return screen.findByTestId('searchContainer');
+      },
+      withinComponent() {
+        return within(screen.getByTestId('searchContainer'));
+      },
+      searchButton() {
+        return this.withinComponent().findByRole('button', { name: 'Hae' });
+      },
+      dateRangeFilter() {
+        return this.withinComponent().findByLabelText('Valitse ajankohta');
+      },
+      dateRangeCheckbox(range: string) {
+        return this.withinComponent().findByRole('checkbox', {
+          name: range,
+        });
+      },
+      neighborhoodFilter() {
+        return this.withinComponent().findByLabelText('Etsi alue');
+      },
+      neighborhoodCheckbox(n: Neighborhood) {
+        return this.withinComponent().findByRole('checkbox', {
+          name: n.name.fi,
+        });
+      },
+      placeFilter() {
+        return this.withinComponent().findByLabelText('Etsi tapahtumapaikka');
+      },
+      placeSearchInput() {
+        return this.withinComponent().findByLabelText('Kirjoita hakusana');
+      },
+      placeCheckbox(place: PlaceFieldsFragment) {
+        return this.withinComponent().findByRole('checkbox', {
+          name: place.name.fi,
+        });
+      },
+      searchInput() {
+        return this.withinComponent().findByPlaceholderText(
+          'Kirjoita hakusana, esim. rock tai jooga'
+        );
+      },
+    };
```

Comment:
To me seems like these don't need to be functions that are called every time you need the element. These could be just functions and that are not nested in selectors object (then we could also avoid using this keyword). "This" keyword forces us to use this api in a specific way to work correctly. These functions can be then wrapped with selectors property in return statemen

Reply:
Like this:

```
    const component = screen.findByTestId('searchContainer');
    const withinComponent = within(screen.getByTestId('searchContainer'));
    const searchButton = withinComponent.findByRole('button', { name: 'Hae' });
    const dateRangeFilter = withinComponent.findByLabelText(
      'Valitse ajankohta'
    );
    const neighborhoodFilter = withinComponent.findByLabelText('Etsi alue');
    const placeFilter = withinComponent.findByLabelText('Etsi tapahtumapaikka');
    const placeSearchInput = withinComponent.findByLabelText(
      'Kirjoita hakusana'
    );
    const searchInput = withinComponent.findByPlaceholderText(
      'Kirjoita hakusana, esim. rock tai jooga'
    );
    const neighborhoodCheckbox = (n: Neighborhood) =>
      withinComponent.findByRole('checkbox', {
        name: n.name.fi,
      });
    const dateRangeCheckbox = (range: string) =>
      withinComponent.findByRole('checkbox', {
        name: range,
      });
    const placeCheckbox = (place: PlaceFieldsFragment) =>
      withinComponent.findByRole('checkbox', {
        name: place.name.fi,
      });
```

Reply:
here is comment below why it does not work